### PR TITLE
Draft of abort_timeout and redis namespace

### PR DESCRIPTION
**Pull request by dp-alvarez**
### Backend
- Keys are now strings
- Keys are stored with a dict containing parameters to the abort request
- Redis takes an optional namespace to store keys on

### Middleware
- Abort includes an abort_timeout parameter
- Tasks can check if an abort is requested via `abort_requested`. Returns the number of seconds the task has before an `Abort` exception is raised or `None` if no abort is requested
- "Main loop" split into fetching abort messages and aborting pending tasks

### TODO
- Gevent abort manager
- Tests
- Update documentation

### CODE REVIEW

File: src/dramatiq_abort/backend.py [hunk 1]
```diff
@@ -1,14 +1,14 @@
...
 
 class EventBackend(abc.ABC):
     """ABC for event backends."""
 
     @abc.abstractmethod
     def wait_many(
-        self, keys: List[bytes], timeout: int
-    ) -> Optional[bytes]:  # pragma: no cover
+        self, keys: Iterable[str], timeout: int
+    ) -> Tuple[Optional[str], Optional[Dict[str, Any]]]:  # pragma: no cover
```
Comment (isra17) [inline src/dramatiq_abort/backend.py:11]: I would rather use `NamedTuple` if the tuple components are not obvious. Also, Let's make the tuple optional, not its component.

```python
class Event(NamedTuple):
  type: str
  params: Dict[str, Any]

def wait_many(...) -> Optional[Event]: ...
def poll(...) -> Optional[Event]: ...
def notify(self, items: Iterable[Event], ...): ...
```
```

File: src/dramatiq_abort/backends/redis.py [hunk 2]
```diff
@@ -14,38 +18,97 @@ class RedisBackend(EventBackend):
     .. _`redis client`: https://pypi.org/project/redis/
     """
 
-    def __init__(self, *, client: Any) -> None:
+    def __init__(self, *, client: Any, namespace: Optional[str] = None) -> None:
```
Comment (isra17) [inline src/dramatiq_abort/backends/redis.py:21]: No need for Optional, just default to `DEFAULT_NAMESPACE`. `str` are immutable, so safe to share as default.

File: src/dramatiq_abort/backends/redis.py [hunk 3]
```diff
@@ -14,38 +18,97 @@ class RedisBackend(EventBackend):
...
+    def notify(self, items: Iterable[Tuple[str, Dict[str, Any]]], ttl: int) -> None:
         with self.client.pipeline() as pipe:
-            for key in keys:
-                pipe.rpush(key, b"x")
+            for key, val in items:
+                pipe.rpush(*self._encode(key=key, value=val))
                 pipe.pexpire(key, ttl)
             pipe.execute()
+
+    def _encode(
```
Comment (isra17) [inline src/dramatiq_abort/backends/redis.py:62]: I would rather split `_encode` into `_encode_key` and `_encode_value` to have:

```python
rpush(self._encode_key(event.key), self._encode_value(event.params))
lpop(self._encode_key(key))
```

File: src/dramatiq_abort/backends/redis.py [hunk 4]
```diff
@@ -1,9 +1,13 @@
-from typing import Any, Iterable, List, Optional, Tuple
+import json
+from typing import Any, Dict, Iterable, Optional, Tuple, Union, overload
 
 import redis
 
 from ..backend import EventBackend
 
+MARKER = "dramatiq-abort_marker"
```
Comment (isra17) [inline src/dramatiq_abort/backends/redis.py:8]: to be fair, I don't think the `MARKER` is worth to keep with namespace. I would remove this.

File: src/dramatiq_abort/backends/redis.py (hunk 3 above)
Comment (isra17) [reply src/dramatiq_abort/backends/redis.py:62]: And do the same for `_decode_*`

File: src/dramatiq_abort/middleware.py [hunk 5]
```diff
@@ -144,53 +144,74 @@ def abort(
...
+            threading.get_ident(), (None, None)
+        )
+        if abort_time is None:
+            return None
+        return 1000 * (abort_time - time.monotonic())
 
-    def _handle(self) -> None:
-        message_ids = list(self.manager.abortables.keys())
+    def _get_abort_requests(self) -> None:
+        message_ids = self.manager.abortable_messages.keys()
```
Comment (isra17) [inline src/dramatiq_abort/middleware.py:168]: why did you remove the `list`? It's required to avoid iterating on mutated .dictionnary

File: src/dramatiq_abort/middleware.py [hunk 6]
```diff
@@ -144,53 +144,74 @@ def abort(
...
+    else:
+        raise RuntimeError("The default broker doesn't have an abortable backend.")
 
 
 def abort(
     message_id: str,
     middleware: Optional[Abortable] = None,
     abort_ttl: Optional[int] = None,
     mode: AbortMode = AbortMode.ABORT,
+    abort_timeout: int = 0,
```
Comment (isra17) [inline src/dramatiq_abort/middleware.py:214]: Please update the docstring

File: src/dramatiq_abort/middleware.py [hunk 7]
```diff
@@ -231,24 +253,54 @@ class _CtypesAbortManager:
...
+                thread_id
+                for thread_id, (_, abort_time) in self.abort_requests.items()
+                if time.monotonic() >= abort_time
+            ]
+
+            for thread_id in toabort:
+                message_id, abort_time = self.abort_requests.pop(thread_id, ("", None))
+                saved_thread_id = self.abortable_messages.pop(message_id, None)
+                assert saved_thread_id == thread_id
+                if thread_id is None or abort_time is None:
```
Comment (isra17) [inline src/dramatiq_abort/middleware.py:295]: I don't see how this condition could ever be `True`. We are inside the lock, so `abort_request[thread_id]` is guaranteed to have value.

Comment (isra17) [reply src/dramatiq_abort/middleware.py:295]: Actually I'm pretty sure that we really only need line 300-303

**Review by isra17 — CHANGES_REQUESTED** [review]
Few comments and improvements! Overall look good to me, thanks for the work!

File: src/dramatiq_abort/middleware.py (hunk 5 above)
Comment (dp-alvarez) [reply src/dramatiq_abort/middleware.py:168]: Nice catch! I had edited that code and forgot the list when I changed it back

File: src/dramatiq_abort/middleware.py (hunk 7 above)
Comment (dp-alvarez) [reply src/dramatiq_abort/middleware.py:295]: That is true, the `if` really should be removed.

Although we technically only need the last few lines I think it is better to keep the pops. If for whatever reason remove_abortable doesn't run we would end up aborting the wrong task, so remove them from the list just in case.

**Comment by dp-alvarez** [comment]
Awesome! I will send in another PR once I finish the docs and tests.

**Comment by dp-alvarez** [comment]
I extracted AbortManager into a abstract class and added support for gevent. Let me know if it looks ok.

Docstrings, tests and documentation still todo.

File: src/dramatiq_abort/abort_manager.py [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+    """
+
+    logger: Any
+    lock: ContextManager[Any]
+    abortable_messages: Dict[str, Any]
+    abort_requests: Dict[Any, Tuple[str, float]]
+
+    @abc.abstractmethod
+    def __init__(self, logger: Optional[Logger] = None):
+        self.logger = logger or get_logger(__name__, type(self))
```
Comment (isra17) [inline src/dramatiq_abort/abort_manager.py:47]: I think we might as well set `abortable_messages` and `abort_requests` here since its quite core to the base.

`lock` could also be defined as an abstract property, forcing implementer to define it, ie:

```
@abc.abstractmethod
@property
def lock(self) -> ContextManager[Any]: ...
  # CTypesManager impl: return self._lock
```

File: src/dramatiq_abort/abort_manager.py [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+    """ABC for raising Abort exceptions in threads.
+
+    :param logger: The logger for abort log lines.
+    :type logger: :class:`logging.Logger`
+    """
+
+    logger: Any
+    lock: ContextManager[Any]
+    abortable_messages: Dict[str, Any]
+    abort_requests: Dict[Any, Tuple[str, float]]
```
Comment (isra17) [inline src/dramatiq_abort/abort_manager.py:43]: Can we also use a `AbortRequest` NamedTuple here?

File: src/dramatiq_abort/abort_manager.py [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+
+    def get_abortables(self) -> List[str]:
+        return list(self.abortable_messages.keys())
+
+    def get_abort_request(self, message_id: Optional[str] = None) -> Optional[float]:
+        if message_id:
+            thread = self.abortable_messages.get(message_id, None)
+        else:
+            thread = self.get_current_thread()
+        _, abort_time = self.abort_requests.get(thread, (None, None))
```
Comment (isra17) [inline src/dramatiq_abort/abort_manager.py:75]: I would rather keep the longer, but simpler alternative:

```python
abort_request = self.abort_requests.get(thread)
if not abort_request:
  return None
return abort_request.abort_time # assuming namedtuple
```

File: src/dramatiq_abort/abort_manager.py [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+    def abort_pending(self) -> None:
+        with self.lock:
+            toabort = [
+                thread
+                for thread, (_, abort_time) in self.abort_requests.items()
+                if time.monotonic() >= abort_time
+            ]
+
+            for thread in toabort:
+                message_id, abort_time = self.abort_requests.pop(thread, ("", None))
```
Comment (isra17) [inline src/dramatiq_abort/abort_manager.py:98]: as above, not a fan of artificial defaults, prefer:

```python
abort_request = self.abort_requests.pop(thread, None)
if abort_request:
  self.abortable_messages.pop(abort_request.key, None)
```

**Comment by dp-alvarez** [comment]
Just finished updating the docs, tests and requested changes (along with some misc fixes along the way).
Appreciate the review! Certainly made the code much cleaner.

Let me know if it needs any tweaks, or feel free to change anything as you see fit.

**Comment by isra17** [comment]
I merged by hand (And fixed the CI), thanks a lot again!

**Comment by dp-alvarez** [comment]
I see a new version was released too, great.

And thank you for this project! Saved me a lot of time compared to doing it from scratch.
